yojimbo: conan 2 support with new premake generator

perseoGI opened the pull request:
### Summary  
Changes to the recipe: **yojimbo/1.2.5**

Main changes:

- The version has been updated from 1.2.1 to 1.2.5, removing the restriction that limited it to the x86_64 target only.
- Unvendor `libsodium` as part of the refactor
- Require our `premake` recipe from Conan Center Index
- Patches:
  - Remove custom `targetdir` which is now managed by `PremakeToolchain`
  - Remove `FatalWarnings` which is an equivalence of `-Werror` (typically removed in CCI) to avoid warning as errors (this project already has warnings)
- A complete refactor of the recipe has been performed by incorporating the `PremakeToolchain`, significantly simplifying the previous version.
- Configure build by defining custom compilation definitions per project avoiding compiling internal tests (thanks @ErniGH)
- Use new `Premake` generator (which was merged in https://github.com/conan-io/conan/pull/17898 and released in Conan v2.18), targeting only the project libraries avoiding compiling any test
- About `package_info`, even though [upstream Premake file](https://github.com/mas-bandwidth/yojimbo/blob/main/premake5.lua#L55-L57) does not explicitly declare any of their internal dependencies:
  - `netcode` (`libnetcode`) depends on `libsodium`: https://github.com/mas-bandwidth/yojimbo/blob/main/netcode/netcode.c#L39
  - `yojimbo` (`libyojimbo`) which is the main library, depends on the rest of the libraries:
    - `libsodium`: https://github.com/mas-bandwidth/yojimbo/blob/main/source/yojimbo.cpp#L32
    - `netcode`: https://github.com/mas-bandwidth/yojimbo/blob/main/source/yojimbo_client.cpp#L5
    - `reliable`: https://github.com/mas-bandwidth/yojimbo/blob/main/source/yojimbo_base_client.cpp#L31
    - `tlsf`: https://github.com/mas-bandwidth/yojimbo/blob/main/source/yojimbo_allocator.cpp#L32

perseoGI committed:
yojimbo: conan 2 support with new premake generator

perseoGI committed:
Fix licence packaging

perseoGI committed:
Split in components

Werni2A commented on recipes/yojimbo/all/conanfile.py:
```diff
@@ -1,117 +1,76 @@
+from conan.tools.premake import Premake, PremakeDeps, PremakeToolchain
+
+
+required_conan_version = ">=2.18.0"
```
```suggestion
required_conan_version = ">=2.19.0"
```
Looks like https://github.com/conan-io/conan/pull/18572 is required to fix this build issue.

`Error: C:/workspace/cci_prod_PR-27827/conan-home/p/b/yojim66f270c3c143d/b/build-release/conanfile.premake5.lua:3: invalid escape sequence near '"C:\w'`

perseoGI replied:
Oh yes thank you, 

I'm aware of it @Werni2A!
I was just trying to fix Linux CI.

perseoGI committed:
Update recipes/yojimbo/all/conanfile.py

Co-authored-by: Dominik Wernberger <dominik.wernberger@gmx.de>

perseoGI committed:
Use version range for premake

perseoGI committed:
Disable internal tests

Co-authored-by: Ernesto de Gracia Herranz <ernestodgh@jfrog.com>

ErniGH reviewed (dismissed):
LGTM

The use of the new toolchain greatly improves the recipe. The description was very helpful. 👍

uilianries commented on recipes/yojimbo/all/conanfile.py:
```diff
@@ -1,117 +1,79 @@
+        copy(self, "LICENCE", self.source_folder, os.path.join(self.package_folder, "licenses"))
+        for folder in ("include", "tlsf", "netcode", "reliable", "serialize"):
+            # Add a namespace to avoid conflicts with other libraries
+            copy(self, "*.h", os.path.join(self.source_folder, folder), os.path.join(self.package_folder, "include", "yojimbo"))
```
```suggestion
        copy(self, "*.h", os.path.join(self.source_folder, folder), os.path.join(self.package_folder, "include")
```

The upstream does not point a namespace, but consuming directly from header: https://github.com/mas-bandwidth/yojimbo/blob/main/test.cpp#L25

uilianries commented on recipes/yojimbo/all/conanfile.py:
```diff
@@ -1,117 +1,79 @@
 
-    def build(self):
+    def build_requirements(self):
+        self.tool_requires("premake/[>=5 <6, include_prerelease]")
```
```suggestion
        # INFO: include pre-release only because all versions are pre-release. 
        self.tool_requires("premake/[>=5 <6, include_prerelease]")
```

uilianries commented on recipes/yojimbo/all/conanfile.py:
```diff
@@ -1,117 +1,79 @@
+        tc = PremakeToolchain(self)
+        # Disable internal tests per project
+        tc.project("netcode").extra_defines = ["NETCODE_ENABLE_TESTS=False"]
+        tc.project("reliable").extra_defines = ["RELIABLE_ENABLE_TESTS=False"]
```
```suggestion
        tc.project("netcode").extra_defines = ["NETCODE_ENABLE_TESTS=0"]
        tc.project("reliable").extra_defines = ["RELIABLE_ENABLE_TESTS=0"]
```

False is not a C++ keyword. You could use false instead, but usually int 0 is more common.

uilianries commented on recipes/yojimbo/all/conanfile.py:
```diff
@@ -1,117 +1,79 @@
+    def package_info(self):
+        # Netcode component
+        self.cpp_info.components["netcode"].libs = ['netcode']
+        self.cpp_info.components["netcode"].requires = ["libsodium::libsodium"]
```
```suggestion
        self.cpp_info.components["netcode"].requires = ["libsodium::libsodium"]
        if self.settings.os == "Windows":
            self.cpp_info.components["netcode"].system_libs = ["Ws2_32", "Iphlpapi"]
```

https://github.com/mas-bandwidth/yojimbo/blob/v1.2.5/netcode/netcode.c#L390

https://github.com/mas-bandwidth/yojimbo/blob/v1.2.5/netcode/netcode.c#L471

uilianries commented on recipes/yojimbo/all/test_package/test_package.cpp:
Let's simplify the test package; it invokes a ton of calls, including WSAStartup and libsodium_init. Let's validate the linkage and keep it simple.

No namespace in the folder, as the upstream does not include one. 

```cpp
#include <iostream>
#include <yojimbo.h>

int main() {
  uint8_t buffer[10] = {0};
  yojimbo_random_bytes(buffer, sizeof(buffer));
  std::cout << "Yojimbo random byte: " << buffer[1] << std::endl;
  return 0;
}
```

uilianries commented on recipes/yojimbo/all/conandata.yml:
```diff
@@ -1,4 +1,4 @@
-  "1.2.1":
-    url: "https://github.com/networkprotocol/yojimbo/archive/refs/tags/v1.2.1.tar.gz"
-    sha256: "293033a8a85ce5d210db6e99da337bc9307a02e8afe5e38afefc694c13359e71"
+  "1.2.5":
```
**NOTE**: Version 1.2.5 is preview (pre-release), not stable one: https://github.com/mas-bandwidth/yojimbo/releases. The latest stable version is 1.2.3

uilianries reviewed (commented):
I have a few comments, otherwise, looking good.

uilianries committed:
Apply review suggestions

https://github.com/conan-io/conan-center-index/pull/27827#pullrequestreview-4465400640

Signed-off-by: Uilian Ries <uilianr@jfrog.com>

uilianries reviewed (approved):
LGTM! 🙌 

#### Reviewer Notes
* Version 1.2.5 is the latest, but it is not stable, is a pre-release/preview. The latest stable is 1.2.3 (2023): https://github.com/mas-bandwidth/yojimbo/releases
* The license remains BSD-3-Clause: https://github.com/mas-bandwidth/yojimbo?tab=BSD-3-Clause-1-ov-file#readme
* Test package uses a utility to keep it as minimal as possible
* libsodium used here is not the latest, but keeps aligned with other recipes in CCI
* Premake packages available in Conan Center are pre-release
* PremakeToolchain requires Conan 2.18, 2.19 has important improvements: https://docs.conan.io/2/changelog.html#id335
* The netcode, reliable, and tlsf are projects from the same organization, but vendorized in Yojimbo because they are mandatory requirements (not explicitly in the Premake, but use headers and methods)
